presto-pinot/src/main/java/io/prestosql/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,433 @@
+import static io.prestosql.pinot.PinotErrorCode.PINOT_DECODE_ERROR;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_EXCEPTION;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_INSUFFICIENT_SERVER_RESPONSE;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_UNEXPECTED_RESPONSE;
+import static io.prestosql.pinot.PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE;
+import static io.prestosql.pinot.PinotMetrics.doWithRetries;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.Long.parseLong;
+import static java.util.Objects.requireNonNull;
+
+public class PinotBrokerPageSource
+        implements ConnectorPageSource
+{
+    private static final Logger log = Logger.get(PinotBrokerPageSource.class);
+
+    private static final String REQUEST_PAYLOAD_TEMPLATE = "{\"pql\" : \"%s\" }";
+    private static final String QUERY_URL_TEMPLATE = "http://%s/query";
+
+    private static final String PINOT_INFINITY = "∞";
+    private static final String PINOT_POSITIVE_INFINITY = "+" + PINOT_INFINITY;
+    private static final String PINOT_NEGATIVE_INFINITY = "-" + PINOT_INFINITY;
+
+    private static final Double PRESTO_INFINITY = Double.POSITIVE_INFINITY;
+    private static final Double PRESTO_NEGATIVE_INFINITY = Double.NEGATIVE_INFINITY;
+
+    private final BrokerPqlWithContext brokerPql;
+    private final List<PinotColumnHandle> columnHandles;
+    private final PinotClusterInfoFetcher clusterInfoFetcher;
+    private final ConnectorSession session;
+    private final ObjectMapper objectMapper;
+
+    private boolean finished;
+    private long readTimeNanos;
+    private long completedBytes;
+
+    public PinotBrokerPageSource(
+            ConnectorSession session,
+            BrokerPqlWithContext brokerPql,
+            List<PinotColumnHandle> columnHandles,
+            PinotClusterInfoFetcher clusterInfoFetcher,
+            ObjectMapper objectMapper)
+    {
+        this.brokerPql = requireNonNull(brokerPql, "broker is null");
+        this.clusterInfoFetcher = requireNonNull(clusterInfoFetcher, "cluster info fetcher is null");
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.session = requireNonNull(session, "session is null");
+        this.objectMapper = requireNonNull(objectMapper, "object mapper is null");
+    }
+
+    private static Double parseDouble(String value)
+    {
+        try {
+            return Double.valueOf(value);
+        }
+        catch (NumberFormatException ne) {
+            switch (value) {
+                case PINOT_INFINITY:
+                case PINOT_POSITIVE_INFINITY:
+                    return PRESTO_INFINITY;
+                case PINOT_NEGATIVE_INFINITY:
+                    return PRESTO_NEGATIVE_INFINITY;
+            }
+            throw new PinotException(PINOT_DECODE_ERROR, Optional.empty(), "Cannot decode double value from pinot " + value, ne);
+        }
+    }
+
+    private void setValue(Type type, BlockBuilder blockBuilder, String value)
```

Comment:
This is not going to be great, performance-wise (due to all the branches, etc). Take a look at how we structured the decoders for the Elasticsearch connector:

https://github.com/prestosql/presto/tree/master/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders
https://github.com/prestosql/presto/blob/master/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchPageSource.java#L187
https://github.com/prestosql/presto/blob/master/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchPageSource.java#L252-L321